src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php

```diff
@@ -778,6 +778,31 @@ public function testSetAllowedTypesFailsIfUnknownOption()
         $this->resolver->setAllowedTypes('foo', 'string');
     }
 
+    public function testResolveTypedWithUnion()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', 'string|int');
+
+        $options = $this->resolver->resolve(['foo' => 1]);
+        $this->assertSame(['foo' => 1], $options);
+
+        $options = $this->resolver->resolve(['foo' => '1']);
+        $this->assertSame(['foo' => '1'], $options);
+    }
+
+
+    public function testResolveTypedWithUnionOfArray()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', '((string|int)[])|((bool|int)[])');
```

Comment:
Isn't it the same as `(string|int|bool)[]`?

Reply:
Technically no.
`[1, 'foo', bool]` is accepted by `(string|int|bool)[]` but not by `((string|int)[])|((bool|int)[])` since the array is not accepted by `(string|int)[]` or `(bool|int)[]`.

This example is not useful since I don't expect someone to really use this exact definition, but I wanted to write a test with some complex expression `(SomeExpressionWithParenthesis|AnotherExpressionWithParenthesis)`, because my first implementation of removing `()`
- changed `((string|int)[])|((bool|int)[])` to `(string|int)[])|((bool|int)[]`
- consider as the array of `(string|int)[])|((bool|int)`
- then, checked either `string|int)[]` or `(bool|int`

So this is mainly a test for the non-enclosing parenthesis

Reply:
I see, so you want an array of `string|int` or an array of `bool|int` but  not a mix of them.

Reply:
`((string|int)[])|((bool|int)[])` why the extra parentheses are necessary in this case? can't it be just `(string|int)[]|(bool|int)[]`?

Reply:
I had an issue about the order the string was processed.

`int|(string|bool)[] was considered as an array of `int|(string|bool)` rather than an union of `int` and `(string|bool)[]`.

But I reworked the implementation to split on `|` before removing `[]` and I think I finally found a way to give `|` a better priority (to use the same syntax than in phpdoc).